Order cycle creation, checkout options, UX improvement

tonyaraujop opened the pull request:
#### What? Why?

- Closes #10441 

#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- Follow the steps to reproduce on #10441 . 

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category: User facing changes | Technical changes

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

tonyaraujop commented:
Hey @RachL This is the draft PR that I mentioned. Please, see the following screenshot to discuss the error message. I'm using an example error message, if you have any suggestion towards this, please let me know. Also, what is the expected style of the error message?

![image](https://user-images.githubusercontent.com/92229784/221610008-2be22db9-90a5-407f-a361-d94bccb1d315.png)

RachL commented:
@jibees what do you think? Given the type of saving bar, I would tend to use

![image](https://user-images.githubusercontent.com/6525576/222484995-251c85be-2c00-4962-b7ce-0239e260da34.png)

jibees commented:
I do agree.
From a dev perspective, I think you should use `StatusMessage.display`

tonyaraujop commented:
Thanks. I'll work on this and submit for review.

tonyaraujop commented:
This is a short video for this fix.

[Screencast from 06-03-2023 10:00:31.webm](https://user-images.githubusercontent.com/92229784/223117592-d96d6531-cbab-4a05-8ffd-e837fdbb8a53.webm)

RachL commented:
@tonyaraujop nice it's looking good :muscle: our team will review your PR soon, thank you for this work!

nolleto commented on app/webpacker/controllers/checkbox_selection_controller.js:
```diff
@@ -0,0 +1,40 @@
+
+  disableButtons() {
+    const groupStates = this.checkboxTargets.reduce((acc, element) => {
+     const group = element.getAttribute("data-group")
```
You can also use `.dataset`

```suggestion
     const { group } = element.dataset
```

nolleto commented on app/webpacker/controllers/checkbox_selection_controller.js:
```diff
@@ -0,0 +1,40 @@
+    const error = this.errorMessageTarget
+    error.style.display = isVisible ? "block" : "none"
+  }
+}
```
Refactor idea: 
```js
import { Controller } from "stimulus";

const isAllFalse = (values) => values.every((isChecked) => !isChecked)

export default class extends Controller {
  static targets = ["checkbox", "errorMessage"]

  get disableElements() {
    return document.querySelectorAll("[data-checkbox-selection-button]")
  }

  get groupStates() {
    return this.checkboxTargets.reduce((acc, element) => {
      const { group } = element.dataset
      const groupValues = acc[group] || []

      return {
        ...acc,
        [group]: [...groupValues, element.checked]
      }
    }, {})
  }

  connect() {
    this.handleError(false)
  }

  disableButtons() {
    const isAllUnchecked = Object.values(this.groupStates).flatMap(isAllFalse).some(Boolean)

    this.disableElements.forEach((element) => {
      element.disabled = isAllUnchecked
    })

    this.handleError(isAllUnchecked)
  }

  handleError(isVisible) {
    const error = this.errorMessageTarget
    error.style.display = isVisible ? "block" : "none"
  }
}
```

mkllnk commented on app/views/admin/order_cycles/checkout_options.html.haml:
```diff
@@ -78,11 +78,11 @@
   %div#save-bar
     %div.container
+      %span.error-message{"data-checkbox-selection-target":"errorMessage"}
+        Please select at least one shipping method and one payment method.
```
Any text should be translated with I18n.

dacook replied:
I think this would be the section to add the message: https://github.com/openfoodfoundation/openfoodnetwork/blob/df8283661d841c83e17f73d3be7b1fd88dc612f2/config/locales/en.yml#L1254-L1265
These messages can be included in the template with the `t()` helper method (as is done elsewhere in this file).

mkllnk reviewed (commented):
Thanks for working on this! Your demonstration video looks really good. That's exactly the outcome we were hoping for.

I will need to take some time another day to look at your code a bit closer though. I'm new to StimulusJS and I'm not sure this is the best solution. @nolleto's suggestion seems to be a good improvement. But I'm also wondering if we complicate it by checking two groups of checkboxes within the same controller explicitly. Could we generalise the controller to handle any group it's attached to and then we invoke it for each group separately? I'm not sure if I'm clear.

For example, if you look at the select-all-controller then it doesn't know about different groups of checkboxes. It's applied twice to each section. Your challenge is that you need to access another element outside the checkbox area but that is possible, too. You could use a fixed DOM selector but StimulusJS may have better solutions. You can then split the error message into two specific ones, each being a target for the each controller instance.

Does this make sense?

dacook reviewed (commented):
Great to have immediate feedback with a client-side fix, thanks!

I'm also new to Stimulus but this looks reasonable to me. I agree that the suggested refactor looks good so would suggest applying that.

I'm not sure about Maikel's has suggestion because both groups still need to interact with the one error message area, so it's probably simpler to keep it structured this way.

mkllnk commented:
@dacook 

> both groups still need to interact with the one error message area

We can have two different error messages though which would be more specific. So each controller instance would interact with its own error message.

tonyaraujop commented:
Thank you for your comments. I'll work on the refactor suggested by @nolleto. I'm also new to Stimulus and I'm using the controller as a "global" one because we need to make sure that whenever a group isn't select we show the error e disable the submit buttons. Hope to submit more change soon. Thank you, everyone.

tonyaraujop commented:
Hello everyone, I added the suggestions mentioned by @nolleto , but I couldn't manipulate the controller with separate instances to access the state of a different group, so I kept the structure I already had and added to the error handling method, the error message, using the group that is unchecked. Here is a new video related with those changes.

[Screencast from 14-03-2023 22:56:35.webm](https://user-images.githubusercontent.com/92229784/225185628-938f1c17-adf4-4e1d-9fe6-f4915c05f5d3.webm)

rioug reviewed (changes requested):
Nice work ! I left a few comments on your test, plus they are currently broken due to your error messages changes I think. 
It's a bit tricky to do testing when it involves translations, but you can see an example here on how to mock `I18n` global object : https://github.com/openfoodfoundation/openfoodnetwork/blob/2f6b730ed4de1fe0d21c5fab5dbea06b38136dbd/spec/javascripts/stimulus/unsaved_changes_controller_test.js#L123-L137

Can we change the name as well ? `checkbox_selection_controller` is a bit vague, what about something like : `checkbox_mandatory_selection` ? What do you think ?

Re @mkllnk comments, one way to address the issue would be pass an error message to the controller, that way the controller doesn't need to know which "group" it is handling.
```
// in the HTML
 "data-checkbox-selection-error-message-value": "your error message"
 
 // in the controller 
 static values = { errorMessage: String }
```
That way you could use it in a similar way to `select_all_controller`, but we might need to move error message, as I am not sure the error message span in the save bar will still be in the scope of the controller. 
I would be in favour of going with this implementation for now, and we could look into making it a bit more generic down the line if the need arise. What do you think @mkllnk ?

tonyaraujop commented:
> Nice work ! I left a few comments on your test, plus they are currently broken due to your error messages changes I think. It's a bit tricky to do testing when it involves translations, but you can see an example here on how to mock `I18n` global object :
> 
> https://github.com/openfoodfoundation/openfoodnetwork/blob/2f6b730ed4de1fe0d21c5fab5dbea06b38136dbd/spec/javascripts/stimulus/unsaved_changes_controller_test.js#L123-L137
> 
> Can we change the name as well ? `checkbox_selection_controller` is a bit vague, what about something like : `checkbox_mandatory_selection` ? What do you think ?
> 
> Re @mkllnk comments, one way to address the issue would be pass an error message to the controller, that way the controller doesn't need to know which "group" it is handling.
> 
> ```
> // in the HTML
>  "data-checkbox-selection-error-message-value": "your error message"
>  
>  // in the controller 
>  static values = { errorMessage: String }
> ```
> 
> That way you could use it in a similar way to `select_all_controller`, but we might need to move error message, as I am not sure the error message span in the save bar will still be in the scope of the controller. I would be in favour of going with this implementation for now, and we could look into making it a bit more generic down the line if the need arise. What do you think @mkllnk ?

Thank you for your review. I'll work on those changes.

sigmundpetersen commented:
Hey @tonyaraujop how are you?👋 
Did you want to continue working on this?
Or we can close it and you can come back to it another time if you like

lin-d-hop commented:
Hi @tonyaraujop Its been a few weeks again and no update. 
Shall we move this out of the active development? You can continue to work on it, we'll just stop chasing you :) And you can ping us to let us know you've submitted a PR so get it back into the dev pipe.

tonyaraujop commented:
Hey @lin-d-hop yeah, please move out from active development. Sorry for this to be open for so long, I'll try to find some time to continue working on this, but in meantime feel free to close this PR.
